realm/realm-library/src/main/cpp/util.hpp

```diff
@@ -58,11 +61,73 @@ JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM *vm, void *reserved);
 #define STRINGIZE_DETAIL(x) #x
 #define STRINGIZE(x) STRINGIZE_DETAIL(x)
 
-// Exception handling
-#define CATCH_STD() \
-    catch (...) { \
-        ConvertException(env, __FILE__, __LINE__); \
+void ConvertException(JNIEnv* env);
+
+// Exception handling using lambdas
+// Credit: http://stackoverflow.com/questions/29202377/how-do-i-force-the-compiler-to-recognize-lambda-type-in-templated-function
+template<typename T, typename F>
+T try_catch(JNIEnv* env, F func) {
+    try {
+        // if an exception already has occurred, it is not possible to call any JNI
+        // functions, and we don't wish to try to execute the JNI function
+        // alternatively, we could clear the exception but then we can easily miss
+        // an error stage
+        if (env->ExceptionCheck() == JNI_FALSE) {
+            // the actual JNI code is executed
+            // it might throw C++ exceptions
+            return func();
+        }
     }
+    catch (std::exception& e) {
+        // we don't wish to do the exception handling in here; this is a templated function, and we might end up
+        // with multiple copies of the code for no good reason
+        ConvertException(env);
+    }
+
+    // when a Java exception is thrown, we return to Java with any return value; as soon as the execution point is
+    // back at the JVM, the Java exception is thrown, and the returned value isn't used for anything
+    return static_cast<T>(0);
+}
+
+// we need to have specialized C++ exceptions so we know what kind of error state has been reached
+// illegal_state - the internal state of the Realm is wrong; it is not serious but execution cannot continue
+class illegal_state : public std::runtime_error {
+public:
+    illegal_state(const std::string& msg) : std::runtime_error(msg) {};
+    illegal_state(const char* msg) : std::runtime_error(msg) {};
+};
+
+// unsupported_operation - the operation is either not supported or implemented; see also java.lang.UnsupportedOperationException
+class unsupported_operation : public std::runtime_error {
+public:
+    unsupported_operation(const std::string& msg) : std::runtime_error(msg) {};
+    unsupported_operation(const char* msg) : std::runtime_error(msg) {};
+};
+
+// fatal_error - the app cannot recover from its error state
+class fatal_error : public std::runtime_error {
+public:
+    fatal_error(const std::string& msg) : std::runtime_error(msg) {};
+    fatal_error(const char* msg) : std::runtime_error(msg) {};
+};
+
+// class_not_found - looking on a class in JVM failed; see also java.lang.ClassNotFoundException
+class class_not_found : public std::runtime_error {
+public:
+    class_not_found(const std::string& msg) : std::runtime_error(msg) {};
+    class_not_found(const char* msg) : std::runtime_error(msg) {};
+};
+
+// null_value - trying to set a required field to null
+class null_value : public std::exception {
```

Comment:
I have no idea ... We have different categories of exception in both java and c++, perhaps stay with `runtime_error` in JNI? Also, see https://github.com/realm/realm-java/pull/3602#pullrequestreview-3622825 , are there any core exception we want them to be thrown in native code without converting for easier core debugging?